contents/java/streams-an-introduction.md

```diff
@@ -0,0 +1,212 @@
+
+When the lambda has a single parameter, the parentheses surrounding the parameter can also be removed,
+```java
+x -> x + 10
+```
+
+For a lambda expression to be compatible with the required functional interface, the lambda expression must have __the same parameters types__ and __a compatible return type__ as the required functional interface.
+
+Although lambda expressions can be expressed in a block, it does not introduce a new level of scoping. Names in the body of the lambda are interpreted in the same way as its enclosing scope, with the addition of the names of its parameters. `this` and `super` can also be used in the lambda body to refer to the enclosing object and the parent class of the enclosing object.
+
+This also means that lambda expressions are able to access local variables of the enclosing scope as well. However, any local variables accessed by a lambda expression must be final or effectively final (ie. cannot be reassigned another value).
+
+To understand more about lambda expressions, take a look at http://www.lambdafaq.org/ ! The website provides easy-to-understand answers to many questions which you may have on lambda expressions. For a detailed use case of lambda expressions, you can read [this Java tutorial](https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html).
+
+
+#### An Example
+
+Now that we have some understanding on functional interface, method references and lambda expressions, let's consolidate our understanding with an example.
+
+`Comparator<T>` may be a familiar interface to you. In previous Java editions, to create an object extending this interface, you could either instantiate the object from a named class or an anonymous class, both of which need to override the abstract methods in `Comparator<T>`. So the code may look something like this:
+```java
+// trying to sort students by height
+students.sort(new Comparator<Student>() {
+    public int compare(Student s1, Student s2) {
+      return Double.compare(s1.getHeight(), s2.getHeight());
+    }
+  });
+```
+Let's face it. Declaring classes (anonymous or not) is quite a pain. Declaring a named class adds to the number of classes you need to maintain while the syntax of a anonymous class is quite an eyesore.
+
+In Java 8, this interface has become a functional interface (surprise, surprise), which allows you to write the code like this instead!
+```java
+// trying to sort students by height
+students.sort((s1, s2) -> Double.compare(s1.getHeight(), s2.getHeight()));
+```
+The compiler is able to infer that a object of type `Comparator<Student>` is expected and that the lambda expression fits into the definition for `compare` (the single abstract method), thus creating an instance of type `Comparator<Student>` with the lambda expression implemented as the `compare` method.
+
+### The Stream API
+
+// source of elements
```

Comment:
I assume you will be adding these by today? Otherwise the TODOs will look weird if merged.